internal/collector/reconcile/gateway.go

```diff
@@ -0,0 +1,139 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/signalfx/splunk-otel-collector-operator/internal/collector"
+	"github.com/signalfx/splunk-otel-collector-operator/internal/naming"
+)
+
+// +kubebuilder:rbac:groups="apps",resources=deployments,verbs=get;list;watch;create;update;patch;delete
+
+// Gateway reconciles the Splunk Otel Gateway required for the instance in the current context.
+func Gateways(ctx context.Context, params Params) error {
+	desired := []appsv1.Deployment{}
+	if params.Instance.Spec.Gateway.Enabled != nil && *params.Instance.Spec.Gateway.Enabled {
+		// TODO(splunk): pass params.Instance.Spec.Gateway instead of params.Instance
+		desired = append(desired, collector.Gateway(params.Log, params.Instance))
+	}
+
+	// first, handle the create/update parts
+	if err := expectedGateways(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the expected deployments: %w", err)
+	}
+
+	// then, delete the extra objects
+	if err := deleteGateways(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the deployments to be deleted: %w", err)
+	}
+
+	return nil
+}
+
+func expectedGateways(ctx context.Context, params Params, expected []appsv1.Deployment) error {
+	for _, obj := range expected {
+		desired := obj
+
+		if err := controllerutil.SetControllerReference(&params.Instance, &desired, params.Scheme); err != nil {
+			return fmt.Errorf("failed to set controller reference: %w", err)
+		}
+
+		existing := &appsv1.Deployment{}
+		nns := types.NamespacedName{Namespace: desired.Namespace, Name: desired.Name}
+		err := params.Client.Get(ctx, nns, existing)
+		if err != nil && k8serrors.IsNotFound(err) {
+			if err = params.Client.Create(ctx, &desired); err != nil {
+				return fmt.Errorf("failed to create: %w", err)
+			}
+			params.Log.V(2).Info("created", "deployment.name", desired.Name, "deployment.namespace", desired.Namespace)
+			continue
+		} else if err != nil {
+			return fmt.Errorf("failed to get: %w", err)
+		}
+
+		// it exists already, merge the two if the end result isn't identical to the existing one
+		updated := existing.DeepCopy()
+		if updated.Annotations == nil {
+			updated.Annotations = map[string]string{}
+		}
+		if updated.Labels == nil {
+			updated.Labels = map[string]string{}
+		}
+
+		updated.Spec = desired.Spec
+		updated.ObjectMeta.OwnerReferences = desired.ObjectMeta.OwnerReferences
+
+		for k, v := range desired.ObjectMeta.Annotations {
+			updated.ObjectMeta.Annotations[k] = v
+		}
+		for k, v := range desired.ObjectMeta.Labels {
+			updated.ObjectMeta.Labels[k] = v
+		}
+
+		patch := client.MergeFrom(existing)
+
+		if err := params.Client.Patch(ctx, updated, patch); err != nil {
+			return fmt.Errorf("failed to apply changes: %w", err)
+		}
+
+		params.Log.V(2).Info("applied", "deployment.name", desired.Name, "deployment.namespace", desired.Namespace)
+	}
+
+	return nil
```

Comment:
Do we want to return early or keep trying and return a multierr?

Reply:
This class is a derivative of an earlier version of [opentelemetry-operator/pkg/collector/reconcile/deployment.go](https://github.com/open-telemetry/opentelemetry-operator/blob/f7aafc50fe0916b359d3070cd3c662002328e24a/pkg/collector/reconcile/deployment.go#L80).

I'd like to leave this for now to match the original code as close as possible to make syncing this Splunk operator code with the Otel operator code easier in the future.